ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -139,205 +119,215 @@ public Driver(QueryState queryState, QueryInfo queryInfo, HiveTxnManager txnMana
+          driverState.executing();
+        } else {
+          String errorMessage = "FAILED: Precompiled query has been cancelled or closed.";
+          CONSOLE.printError(errorMessage);
+          throw DriverUtils.createProcessorException(driverContext, 12, errorMessage, null, null);
+        }
+      } else {
+        driverState.compiling();
+      }
+    } finally {
+      driverState.unlock();
+    }
   }
 
-  /**
-   * @return The current FetchTask associated with the Driver's plan, if any.
-   */
-  @Override
-  public FetchTask getFetchTask() {
-    return driverContext.getFetchTask();
+  private void runPreDriverHooks(HiveDriverRunHookContext hookContext) throws CommandProcessorException {
+    try {
+      driverContext.getHookRunner().runPreDriverHooks(hookContext);
+    } catch (Exception e) {
+      String errorMessage = "FAILED: Hive Internal Error: " + Utilities.getNameMessage(e);
+      CONSOLE.printError(errorMessage + "\n" + StringUtils.stringifyException(e));
+      throw DriverUtils.createProcessorException(driverContext, 12, errorMessage,
+          ErrorMsg.findSQLState(e.getMessage()), e);
+    }
   }
 
-  public void releaseLocksAndCommitOrRollback(boolean commit) throws LockException {
-    releaseLocksAndCommitOrRollback(commit, driverContext.getTxnManager());
-  }
+  public void lockAndRespond() throws CommandProcessorException {
+    // Assumes the query has already been compiled
+    if (driverContext.getPlan() == null) {
+      throw new IllegalStateException(
+          "No previously compiled query for driver - queryId=" + driverContext.getQueryState().getQueryId());
+    }
 
-  /**
-   * @param commit if there is an open transaction and if true, commit,
-   *               if false rollback.  If there is no open transaction this parameter is ignored.
-   * @param txnManager an optional existing transaction manager retrieved earlier from the session
-   *
-   **/
-  @VisibleForTesting
-  public void releaseLocksAndCommitOrRollback(boolean commit, HiveTxnManager txnManager) throws LockException {
-    driverTxnHandler.releaseLocksAndCommitOrRollback(commit, txnManager);
+    try {
+      driverTxnHandler.acquireLocksIfNeeded();
+    } catch (CommandProcessorException cpe) {
+      driverTxnHandler.rollback(cpe);
+      throw cpe;
+    }
   }
 
-  /**
-   * Release some resources after a query is executed
-   * while keeping the result around.
-   */
-  public void releaseResources() {
-    releasePlan();
-    releaseTaskQueue();
+  private void validateTxnListState() throws CommandProcessorException {
+    try {
+      if (!driverTxnHandler.isValidTxnListState()) {
+        LOG.warn("Reexecuting after acquiring locks, since snapshot was outdated.");
+        // Snapshot was outdated when locks were acquired, hence regenerate context,
+        // txn list and retry (see ReExecutionRetryLockPlugin)
+        try {
+          driverTxnHandler.releaseLocksAndCommitOrRollback(false);
```

Comment:
This will collide @mustafaiman's concurrent pull request. You might have to rebase this change